### DISCUSSION TITLE
CPU overload and instability when running 3.1.0 with .compact files from earlier versions.

### DISCUSSION THREAD

**Discussion by kathy1121** [apache/couchdb#2989] [Q&A] [answered]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )
Couchdb consumes 95% cpus event though we extend our disk space.
## Description

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )
We upgrade couchdb to the latest version but it consumes thousands of cpus every moment.
The highest process is beam.smp.
We have many indexers in active tasks.

## Steps to Reproduce

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

[NOTE]: # ( Tell us what you expected to happen. )
Suggest how to optimize the performance and reduce the cpus.
## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )

## Additional Context

[TIP]:  # ( Add any other context about the problem here. )
It consumes lots cpus in our prod environment. Please help solve this issue. Thanks.

**Comment by janl** [comment]
To diagnose this issue, we need to know a lot more details. Can you explain:

-  your setup
- how CouchDB is used by your application(s)
- what your traffic looks like?
- share any error messages from your logs

**Reply by kathy1121** [reply]
-     your setup
cluster
-     how CouchDB is used by your application(s)
not sure.
-     what your traffic looks like?
For example,
Error running query. Reason: {shutdown}
for this error, we extended disk space.
-     share any error messages from your logs
some error:
[error] 2020-07-07T12:14:46.934357Z couchdb@couchdb-couchdb-0.local <0.6155.2141> a963bcbe40 rexi_server: from: couchdb@couchdb-couchdb-0.couchdb-couchdb.couchdb.svc.clusterb.local(<0.29544.2141>) mfa: fabric_rpc:update_docs/3 exit:{function_clause,[{couch_emsort,set_options,[{ems,<0.15003.2141>,undefined,10,100,0,0},{[4520775400,4520473343,4520157749,4519788284,4519479624,4519131193],4519479536}],[{file,"src/couch_emsort.erl"},{line,157}]},{couch_emsort,open,2,[{file,"src/couch_emsort.erl"},{line,154}]},{couch_bt_engine_compactor,bind_emsort,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,634}]},{couch_bt_engine_compactor,open_compaction_files,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,109}]},{couch_bt_engine_compactor,start,4,[{file,"src/couch_bt_engine_compactor.erl"},{line,62}]}]} [{couch_db,collect_results,3,[{file,"src/couch_db.erl"},{line,1266}]},{couch_db,collect_results_with_metrics,3,[{file,"src/couch_db.erl"},{line,1248}]},{couch_db,write_and_commit,4,[{file,"src/couch_db.erl"},{line,1276}]},{couch_db,update_docs,4,[{file,"src/couch_db.erl"},{line,1222}]},{fabric_rpc,with_db,3,[{file,"src/fabric_rpc.erl"},{line,330}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,138}]}]
[error] 2020-07-07T12:14:47.644742Z couchdb@couchdb-couchdb-0.local emulator -------- Error in process <0.27350.2142> on node 'couchdb@couchdb-couchdb-0.couchdb-couchdb.couchdb.svc.clusterb.local' with exit value:
[error] 2020-07-07T12:14:48.062725Z couchdb@couchdb-couchdb-0.local <0.28847.2135> -------- OS Process Error <0.29415.2133> :: {reduce_overflow_error,<<"Reduce output must shrink more rapidly: input size: 26180 output size: 26027">>}

**Reply by kathy1121** [reply]
the beam.smp always consumes lots of cpus. 
the cpus always beyond 5000m

**Reply by janl** [reply]
You have this in the log snippet you pasted:

> Reduce output must shrink more rapidly: input size: 26180 output size: 

This means you have custom reduce function that does not work correctly. There might be other things going wrong, but that is a good first step to fix. https://docs.couchdb.org/en/stable/ddocs/views/intro.html#reduce-rereduce has more details

**Reply by kathy1121** [reply]
Thanks Jan.
We also have this kind of errors.... Are those all related to the cpu issue?
[error] 2020-07-07T13:57:46.567080Z couchdb@couchdb-couchdb-1.local <0.496.2231> -------- CRASH REPORT Process  (<0.496.2231>) with 2 neighbors crashed with reason: no match of right hand value {'EXIT',{function_clause,[{couch_emsort,set_options,[{ems,<0.2085.2230>,undefined,10,100,0,0},{[87879859,87878122,87539677,87149300,86865451],86871490}],[{file,"src/couch_emsort.erl"},{line,157}]},{couch_emsort,open,2,[{file,"src/couch_emsort.erl"},{line,154}]},{couch_bt_engine_compactor,bind_emsort,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,634}]},{couch_bt_engine_compactor,open_compaction_files,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,109}]},{couch_bt_engine_compactor,start,...}]}} at couch_file:pread_binary/2(line:172) <= couch_file:pread_term/2(line:160) <= couch_btree:get_node/2(line:435) <= couch_btree:lookup/3(line:286) <= couch_btree:lookup_kpnode/5(line:306) <= couch_btree:lookup/2(line:276) <= couch_bt_engine:open_local_docs/2(line:407) <= couch_db:open_doc_int/3(line:1653); initial_call: {couch_index,init,['Argument__1']}, ancestors: [<0.15095.2229>], message_queue_len: 0, messages: [], links: [<0.15095.2229>,<0.18012.2229>], dictionary: [], trap_exit: false, status: running, heap_size: 2586, stack_size: 27, reductions: 4035

[error] 2020-07-07T13:57:46.569096Z couchdb@couchdb-couchdb-1.local <0.4336.2214> -------- Supervisor couch_secondary_services had child index_server started with couch_index_server:start_link() at <0.10649.2228> exit with reason no match of right hand value {'EXIT',{function_clause,[{couch_emsort,set_options,[{ems,<0.2085.2230>,undefined,10,100,0,0},{[87879859,87878122,87539677,87149300,86865451],86871490}],[{file,"src/couch_emsort.erl"},{line,157}]},{couch_emsort,open,2,[{file,"src/couch_emsort.erl"},{line,154}]},{couch_bt_engine_compactor,bind_emsort,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,634}]},{couch_bt_engine_compactor,open_compaction_files,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,109}]},{couch_bt_engine_compactor,start,...}]}} at couch_file:pread_binary/2(line:172) <= couch_file:pread_term/2(line:160) <= couch_btree:get_node/2(line:435) <= couch_btree:lookup/3(line:286) <= couch_btree:lookup_kpnode/5(line:306) <= couch_btree:lookup/2(line:276) <= couch_bt_engine:open_local_docs/2(line:407) <= couch_db:open_doc_int/3(line:1653) in context child_terminated

[error] 2020-07-07T13:57:54.958632Z couchdb@couchdb-couchdb-1.clusterb.local <0.23458.2227> -------- CRASH REPORT Process  (<0.23458.2227>) with 0 neighbors exited with reason: {timeout,{gen_server,call,[<0.26178.2225>,{compaction_failed,{{badmatch,{'EXIT',normal}},[{couch_file,pread_binary,2,[{file,"src/couch_file.erl"},{line,172}]},{couch_file,pread_term,2,[{file,"src/couch_file.erl"},{line,160}]},{couch_btree,get_node,2,[{file,"src/couch_btree.erl"},{line,435}]},{couch_btree,stream_node,7,[{file,"src/couch_btree.erl"},{line,764}]},{couch_btree,stream_kp_node,7,[{file,"src/couch_btree.erl"},{line,778}]},{couch_btree,fold,4,[{file,"src/couch_btree.erl"},{line,221}]},...]}}]}} at gen_server:call/2(line:206) <= couch_index_compactor:handle_info/2(line:90) <= gen_server:try_dispatch/4(line:616) <= gen_server:handle_msg/6(line:686) <= proc_lib:init_p_do_apply/3(line:247); initial_call: {couch_index_compactor,init,['Argument__1']}, ancestors: [<0.26178.2225>,<0.19752.2227>], message_queue_len: 1, messages: [{'EXIT',<0.26178.2225>,shutdown}], links: [], dictionary: [], trap_exit: true, status: running, heap_size: 1598, stack_size: 27, reductions: 325

[error] 2020-07-07T13:57:57.629248Z couchdb@couchdb-couchdb-1.local <0.2272.2229> 5b40b22087 rexi_server: from: couchdb@couchdb-couchdb-2.svc.clusterb.local(<15346.6846.2144>) mfa: fabric_rpc:map_view/5 throw:shutdown [{couch_mrview_util,get_view_index_state,5,[{file,"src/couch_mrview_util.erl"},{line,138}]},{couch_mrview_util,get_view,4,[{file,"src/couch_mrview_util.erl"},{line,82}]},{couch_mrview,query_view,6,[{file,"src/couch_mrview.erl"},{line,262}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,138}]}]

**Reply by kathy1121** [reply]
> Reduce output must shrink more rapidly

Regarding this in the couchdb docs:
"CouchDB will be able to compute the final result, but only for views with a few rows. Anything larger will experience a ridiculously slow view build time. To help with that, CouchDB since version 0.10.0 will throw an error if your reduce function does not reduce its input values."

Is the version 0.10.0 will throw error? If the reduce is too slow, can I set the `rereduce = false`?
But after upgrade we don't change any configuration in the couchdb, why leads to the issue?

Thanks a lot.

**Reply by janl** [reply]
you can not set rereduce to anything, CouchDB uses that to tell you that it is in a rereduce phase. Can you share your reduce function here? The issue is still that your reduce function is not written correctly.

**Reply by janl** [reply]
my first guess is that the emsort errors are just follow-up errors from this, so let’s get that resolved first and then see if any other errors remain

**Reply by kathy1121** [reply]
thanks Jan. But I need to check the issue with my teams in next working day. I will reply to you tomorrow.
Thanks a lot!

**Reply by kathy1121** [reply]
hi Jan,

How about this kind of errors?
- CRASH REPORT Process couch_index_server (<0.18289.2787>) with 1 neighbors exited with reason: no match of right hand value {'EXIT',{function_clause,[{couch_emsort,set_options,[{ems,<0.8184.2786>,undefined,10,100,0,0},{[87879859,87878122,87539677,87149300,86865451],86871490}],[{file,"src/couch_emsort.erl"},{line,157}]},{couch_emsort,open,2,[{file,"src/couch_emsort.erl"},{line,154}]},{couch_bt_engine_compactor,bind_emsort,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,634}]},{couch_bt_engine_compactor,open_compaction_files,3,[{file,"src/couch_bt_engine_compactor.erl"},{line,109}]},{couch_bt_engine_compactor,start,...}]}} at couch_file:pread_binary/2(line:172) <= couch_file:pread_term/2(line:160) <= couch_btree:get_node/2(line:435) <= couch_btree:lookup/3(line:286) <= couch_btree:lookup/2(line:276) <= couch_bt_engine:open_local_docs/2(line:407) <= couch_db:open_doc_int/3(line:1653) <= couch_db:open_doc/3(line:289) at couch_index_server:handle_info/2(line:190) <= gen_server:try_dispatch/4(line:616) <= gen_server:handle_msg/6(line:686) <= proc_lib:init_p_do_apply/3(line:247); initial_call: {couch_index_server,init,['Argument__1']}, ancestors: [couch_secondary_services,couch_sup,<0.225.0>], message_queue_len: 0, messages: [], links: [<0.3219.2779>,<0.24173.2783>], dictionary: [], trap_exit: true, status: running, heap_size: 2586, stack_size: 27, reductions: 4345

mfa: fabric_rpc:map_view/5 throw:shutdown [{couch_mrview_util,get_view_index_state,5,[{file,"src/couch_mrview_util.erl"},{line,138}]},{couch_mrview_util,get_view,4,[{file,"src/couch_mrview_util.erl"},{line,82}]},{couch_mrview,query_view,6,[{file,"src/couch_mrview.erl"},{line,262}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,138}]}]

What are they mean?

**Reply by kathy1121** [reply]
> 
> 
> you can not set rereduce to anything, CouchDB uses that to tell you that it is in a rereduce phase. Can you share your reduce function here? The issue is still that your reduce function is not written correctly.

We just upgarded the couchdb from old version to laterst version. But before it worked well without issue.
Why the new version consumes lots of cpus?And the couchdb eats lots of disks every moment.

We also have this kind of errors: 
`mfa: fabric_rpc:map_view/5 throw:shutdown`
`emulator -------- Error in process`
`{{badmatch,{error,{noproc,{gen_server,call`

Could you please help on this?

**Reply by kathy1121** [reply]
> 
> 
> my first guess is that the emsort errors are just follow-up errors from this, so let’s get that resolved first and then see if any other errors remain

Hi Jan,

checked with team, we have quite a few design docs which have custom reduce function. 
But in the earlier version we didn't met this issue. But now the new couchdb eats lots of disk space every moment.
Could you please give some suggestion?

**Reply by wohali** [reply]
If you check the output of `ps` or Process Explorer (Windows), do you see lots of `couchjs` processes? Are they eating all the CPU?

CouchDB 3.1 will be building every view on every database automatically for you now, even ones you never accessed before. This is new functionality in CouchDB 3.x. Perhaps it is building views that were never used before, including views with errors.

You can configure how this auto-building process works, it's [fully documented](https://docs.couchdb.org/en/latest/config/indexbuilds.html). You may want to disable some views from auto-building (by adding `autoupdate: false` to their ddocs), or disable `ken` entirely by setting `[ken] batch_channels = 0` and `[ken] incremental_channels = 0 ` in your `etc/local.ini` file.

**Reply by kathy1121** [reply]
> 
> 
> If you check the output of `ps` or Process Explorer (Windows), do you see lots of `couchjs` processes? Are they eating all the CPU?
> 
> CouchDB 3.1 will be building every view on every database automatically for you now, even ones you never accessed before. This is new functionality in CouchDB 3.x. Perhaps it is building views that were never used before, including views with errors.
> 
> You can configure how this auto-building process works, it's [fully documented](https://docs.couchdb.org/en/latest/config/indexbuilds.html). You may want to disable some views from auto-building (by adding `autoupdate: false` to their ddocs), or disable `ken` entirely by setting `[ken] batch_channels = 0` and `[ken] incremental_channels = 0 ` in your `etc/local.ini` file.

Thank you for your reply.

Yes I checked the `couch.js` process and it doesn't eat all CPU.
Regarding the `ken` configuration, I checked but the result is `{}`   Does it mean we don't have the related configure?

Besides, we found that another core issue is the new version consumes a mess of disks from 40G to 220G in several days. But the old version is stable. Why now eats lots of disks?

Could you please give some advice?

**Reply by kathy1121** [reply]
Hi Team,

regard this case, https://github.com/apache/couchdb/issues/2941
I found that maybe my errors are similar to him. Also our db is very large and now compact failed.

How to solve this issue?

**Comment by kathy1121** [comment, marked as answer]
Refering to https://github.com/apache/couchdb/issues/2941, we also delete old .compact files. The disk usage decrease.
But now we still face issue in a pod:
[error] CRASH REPORT Process
[error]rexi_server:   mfa: fabric_rpc:map_view/5 throw:shutdown
[error]gen_server <0.17218.33> terminated with reason: bad return value {read_beyond_eof,"./data/shards

**Reply by janl** [reply]
Perfect!

**Reply by kathy1121** [reply]
but we still have error in a pod…  one pod still consumes a lot
